rust/lancedb/src/database/listing.rs

```diff
@@ -615,66 +1152,132 @@ impl Database for ListingDatabase {
-                    .map(|e| e == LANCE_EXTENSION);
-                is_lance.unwrap_or(false)
-            })
-            .filter_map(|p| p.file_stem().and_then(|s| s.to_str().map(String::from)))
-            .collect::<Vec<String>>();
-        f.sort();
-        if let Some(start_after) = request.start_after {
-            let index = f
+
+        // When both manifest and directory listing are enabled, we need to deduplicate
+        // by location to avoid listing the same table twice
+        let mut f = if self.manifest_enabled && self.dir_listing_enabled {
+            let mut manifest_request = request.clone();
+            manifest_request.limit = None;
+            manifest_request.start_after = None;
+            let manifest_tables = self.manifest_db.list_tables(manifest_request).await?;
+
+            // Collect directory names from manifest
+            let manifest_locations: HashSet<String> = manifest_tables
                 .iter()
-                .position(|name| name.as_str() > start_after.as_str())
-                .unwrap_or(f.len());
-            f.drain(0..index);
-        }
-        if let Some(limit) = request.limit {
-            f.truncate(limit as usize);
-        }
+                .map(|table_info| table_info.location.clone())
+                .collect();
+
+            // Get table names from manifest
+            let mut table_names: Vec<String> = manifest_tables
+                .iter()
+                .map(|table_info| table_info.name.clone())
+                .collect();
+
+            // Add directory tables that aren't already in the manifest
+            let dir_tables = self.list_directory_tables().await?;
+            for table_name in dir_tables {
+                let dir_name = format!("{}.{}", table_name, LANCE_FILE_EXTENSION);
+                if !manifest_locations.contains(&dir_name) {
+                    table_names.push(table_name);
+                }
+            }
+
+            table_names
+        } else if self.manifest_enabled {
+            let mut manifest_request = request.clone();
+            manifest_request.limit = None;
+            manifest_request.start_after = None;
+            self.manifest_db.table_names(manifest_request).await?
+        } else if self.dir_listing_enabled {
+            self.list_directory_tables().await?
+        } else {
+            Vec::new()
+        };
+
+        apply_pagination(&mut f, request.start_after, request.limit);
         Ok(f)
     }
 
     async fn create_table(&self, request: CreateTableRequest) -> Result<Arc<dyn BaseTable>> {
+        if self.manifest_enabled {
+            return match self.manifest_db.create_table(request).await {
+                Ok(table) => Ok(table),
+                Err(Error::Manifest { table, message }) => {
+                    return if let Some(table) = table {
+                        log::warn!("Failure in manifest operation: {message}");
+                        Ok(table)
+                    } else {
```

Comment:
Can you comment what is going on here?  Why would the manifest_db return an error but we return ok?